wallet-new/src/Cardano/Wallet/API/V1/LegacyHandlers/Transactions.hs

```diff
@@ -67,6 +67,134 @@ newTransaction submitTx Payment {..} = do
+-- 'Word64' representing the number of Lovelace coins, but the operations
+-- require rational numbers. This implies the potential for rounding issues.
+-- Rounding issues give us three options: 'round', 'floor', and 'ceiling'.
+--
+-- Let's consider the strategies with the following data:
+--
+-- >>> outputs = [5, 8, 10]
+-- >>> total = sum outputs
+-- >>> fee = 3
+-- >>> withProportions xs = map (\x -> (x, fromIntegral x / fromIntegral (sum xs)))
+-- >>> withProportions outputs
+-- [(5,0.21739130434782608),(8,0.34782608695652173),(10,0.43478260869565216)]
+-- >>> subtractProportion output = fromIntegral output - (fee * (output / total))
+-- >>> let precise = map (\o -> fromIntegral o - (fee * (o / total))) outputs
+-- >>> precise
+-- [4.875,7.8,9.75]
+--
+-- Our initial total spending was 23, and we want to deduct a fee of 3 from this
+-- amount, for a total disbursement of 20 to these outputs. Let's observe the
+-- result of the different options:
+--
+-- >>> sum (map round precise)
+-- 23
+--
+-- Well, 'round' appears to be precisely what we want! The total lines up
+-- exactly. Is this by chance? We can write a QuickCheck property to test it.
+newTransactionFeesIncluded
+    :: forall ctx m . (V0.MonadWalletTxFull ctx m)
+    => (TxAux -> m Bool) -> Payment -> m (WalletResponse Transaction)
+newTransactionFeesIncluded submitTx pmt = do
+    EstimatedFees (V1 coin) <- wrData <$> estimateFees pmt
+    let outputs = length (pmtDestinations pmt)
+        coinPerOutput = Core.getCoin coin `div` fromIntegral outputs
+        newPayment = pmt
+            { pmtDestinations =
+                map subtractAmount (pmtDestinations pmt)
+            }
+        subtractAmount pmtDistribution = pmtDistribution
+            { pdAmount = V1 $
+                (unV1 (pdAmount pmtDistribution))
+                    `Core.unsafeSubCoin`
+                        Core.mkCoin coinPerOutput
+            }
+
+    newTransaction submitTx newPayment
+
+-- | This function implements the fee distribution logic described in the
+-- comments of 'newTransactionFeesIncluded'. There are two preconditions, and
+-- this function returns 'Nothing' if either are violated:
+--
+-- 1. The fee amount must be less than the total of the outputs. If this is
+--    violated, then it would be impossible for the fee to come entirely from
+--    the outputs, so the transaction is invalid.
+-- 2. None of the outputs may be 0.
+distributeFeesInternal
+    :: Word64 -- ^ The fee amount
+    -> NonEmpty Word64 -- ^ The outputs of the 'PaymentDistribution'
+    -> Maybe (NonEmpty Word64) -- ^ The resulting outputs after subtracting the fee.
+distributeFeesInternal fee outputs
+    | 0 `elem` outputs = Nothing
+    | fromIntegral fee >= total = Nothing
+    | 0 `elem` result = Nothing
+    | otherwise = Just result
+  where
+    result =
+        map (round . subtractFee) outputs
+    total :: Rational
+    total =
+        sum (map fromIntegral outputs)
```

Comment:
there's a huge difference between `fromIntegral (sum outputs)` and `sum (map fromIntegral outputs)`, especially with types that overflow -_-